src/libstore/references.cc

```diff
@@ -81,6 +81,18 @@ StorePathSet scanForReferences(
     Sink & toTee,
     const Path & path,
     const StorePathSet & refs)
+{
+    auto fun = [&](Sink & refsSink) {
+        /* Look for the hashes in the NAR dump of the path. */
+        TeeSink sink { refsSink, toTee };
+        dumpPath(path, sink);
+    };
+    return scanForReferences(fun, refs);
+}
+
+StorePathSet scanForReferences(
+    std::function<void(Sink &)> fun,
```

Comment:
```suggestion
    Source & source,
```
I feel like this should make it less awkward? Note that there are various source to sink and sink to source functions.

Reply:
Yes, I noticed `sinkToSource` (see OP), but I feel like it's a shame to use coroutines for something so simple...

Reply:
Actually, what we really want is a `PathRefScanSink`

Reply:
That probably makes more sense

Reply:
Ugh, this is apparently too much for C++. Tried `PathRefScanSink : public RefScanSink`, but it doesn't seem possible to do the preprocessing (for each path, add its hash to `hashes` and insert it into the `backMap`) *before* calling the parent constructor. I'd like to avoid having an inner `RefScanSink * sink;` member and wiring it up explicitly...

I guess I'll have to make a "smart constructor" function that does the preprocessing and calls the constructor.

Reply:
@ncfavier Here's a trick, you can make static method (or just free-floating function) which the constructor calls!

Reply:
But this needs to happen in the member initialiser list, so we can't feed one output of the function (hashes) to the base constructor and the other output (backMap) to initialise a member...

Reply:
@ncfavier I was thinking about this, see https://stackoverflow.com/questions/4162021/is-it-ok-to-call-a-function-in-constructor-initializer-list I think we can make a blank back map, then make the other one set by mutating the first one.

Reply:
@ncfavier I replied on matrix with details.

Reply:
Made the constructor private and inlined `scanStringForReferences` as requested

Reply:
Thanks!